### ISSUE TITLE
cloudfront invalidation error

### ISSUE THREAD

**Issue by levelos** [laurilehmijoki/s3_website#63]
First, just want to say thanks for such a great library. We just launched our new Jekyll site on S3 and it's been a big help. Occasionally when running `s3_website push` I get the following cloudfront error:

> AWS API call failed. Reason:
> 
> ``` xml
> <?xml version="1.0"?>
> <ErrorResponse xmlns="http://cloudfront.amazonaws.com/doc/2012-05-05/">
> <Error>
> <Type>Sender</Type>
> <Code>InvalidArgument</Code>
> <Message>Your request contains one or more invalid invalidation paths.</Message>
> </Error>
> <RequestId></RequestId>
> </ErrorResponse> 
> ```
> 
> (RuntimeError)

Any thoughts?

**Comment by tkrotoff** [comment]
I still have this problem using v2.8.3 on OS X 10.10:

```
[info] Deploying _site/* to osteo15.com
[succ] Updated sitemap.xml (max-age=864000 | application/xml | gzip)
[succ] Updated 2013/11/22/Interview-sur-l'osteopathie-et-les-femmes-enceintes/index.html (max-age=300 | text/html; charset=utf-8 | gzip)
[succ] Updated tags/index.html (max-age=300 | text/html; charset=utf-8 | gzip)
[succ] Updated feed.xml (max-age=864000 | application/rss+xml | gzip)
[fail] Failed to invalidate the CloudFront distribution (Your request contains one or more invalid invalidation paths. (Service: AmazonCloudFront; Status Code: 400; Error Code: InvalidArgument; Request ID: 72171ed8-bc2f-11e4-b56b-83215c24fdb7))
[info] Summary: Updated 4 files. Transferred 44.5 kB, 144.0 kB/s.
[fail] Failed to push the website to ...
```

If I modify a file that does not contain `'` in its name + push, then no problem:
```
[info] Deploying _site/* to osteo15.com
[succ] Updated sitemap.xml (max-age=864000 | application/xml | gzip)
[succ] Updated 2014/02/17/Medicaments,-grossesse-et-allaitement/index.html (max-age=300 | text/html; charset=utf-8 | gzip)
[succ] Updated feed.xml (max-age=864000 | application/rss+xml | gzip)
[succ] Invalidated 4 items on CloudFront
[info] Summary: Updated 3 files. Transferred 35.3 kB, 108.0 kB/s.
[info] Successfully pushed the website to ...
```

Another test by modifying another file containing a `'` + push:
```
[info] Deploying _site/* to osteo15.com
[succ] Updated sitemap.xml (max-age=864000 | application/xml | gzip)
[succ] Updated 2014/02/19/Questions-reponses-sur-l'hypnose-avec-Delphine-Bourdet/index.html (max-age=300 | text/html; charset=utf-8 | gzip)
[succ] Updated feed.xml (max-age=864000 | application/rss+xml | gzip)
[fail] Failed to invalidate the CloudFront distribution (Your request contains one or more invalid invalidation paths. (Service: AmazonCloudFront; Status Code: 400; Error Code: InvalidArgument; Request ID: e0919706-bc2f-11e4-b56b-83215c24fdb7))
[info] Summary: Updated 3 files. Transferred 38.2 kB, 123.0 kB/s.
[fail] Failed to push the website to ...
```

FYI
- List of valid URL characters: [`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;=`](http://stackoverflow.com/a/1547940/990356)
- List of invalid filename characters under Windows: `\/:*?"<>|` (I think UNIX has less limitations).

**Comment by laurilehmijoki** [comment]
Hi @tkrotoff, I just released the version **2.8.4**, which URL-encodes the `'` character.

Even though `'` is a valid URL character, CloudFront does not accept it unless it's URL-encoded. I don't know why this is the case.

Here's some [documentation](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/cloudfront_2012_03_15/model/InvalidationBatch.html#InvalidationBatch(java.util.List,%20java.lang.String)) on the requirements of a valid invalidation path.

**Comment by tkrotoff** [comment]
@laurilehmijoki tested, works great :+1:

**Comment by laurilehmijoki** [comment]
Thanks for testing, @tkrotoff!

**Comment by mobidev111** [comment]
@laurilehmijoki I get the InvalidArgument error with 2.8.4 when there are filenames with special characters: ä ö ü

**Comment by laurilehmijoki** [comment]
@mobidev111 try upgrading to **2.8.5** – does the problem disappear?

I made a [specific rule](https://github.com/laurilehmijoki/s3_website/commit/009dbf883842eb09ea7f1ba377e9fbbe621fb7dd) for the chars `(ä|ö|ü)`, because I didn't have time to write the general rule. 

Can someone help in defining the general rule?

**Comment by akshaykarle** [comment]
@laurilehmijoki I've attempted to generalize this by escaping all non-ascii characters in #190. I feel this should handle most of the cases. Can you take a look?

**Comment by laurilehmijoki** [comment]
Thanks @akshaykarle! Your pull request is released in version **2.12.2**.
